### PR TITLE
Channel scroll

### PR DESCRIPTION
**Pull request by AndrewMicallef**
Actually fixed the errors. Tested on Ubuntu 14.04 with a 50 channel file. works like a charm!
See issue #39

### CODE REVIEW

File: src/stimfit/gui/graph.cpp [hunk 1]
```diff
@@ -2152,6 +2163,68 @@ void wxStfGraph::OnDown() {
     Refresh();
 }
 
+void wxStfGraph::ChanUp() {
+    // on Control + Up press, increase the channel number by one and
+    // refresh the window
+    
+    int reference_ch = Doc()->GetSecChIndex();  
+    
+    int channel = Doc()->GetCurChIndex() + 1;
```
Comment (neurodroid) [inline src/stimfit/gui/graph.cpp]: Why is there a +1 here but not in line 2170?

Comment (AndrewMicallef) [inline src/stimfit/gui/graph.cpp]: The `reference_ch` variable is just being used to store the value of the second channel. The scrolling shifts the first channel leaving the reference channel in place.

**Comment by AndrewMicallef** [comment]
How is this?

File: src/stimfit/gui/graph.cpp [hunk 2]
```diff
@@ -2152,6 +2163,55 @@ void wxStfGraph::OnDown() {
...
+    */
+    
+    int reference_ch = Doc()->GetSecChIndex();  
+    
+    int new_channel = Doc()->GetCurChIndex() + direction;
+    
+    
+    if (new_channel == reference_ch) {
+        	// Skip the reference channel
+        	channel += units;
```
Comment (neurodroid) [inline src/stimfit/gui/graph.cpp]: Where are `channel` and `units` defined?

File: src/stimfit/gui/graph.cpp [hunk 3]
```diff
@@ -2152,6 +2163,55 @@ void wxStfGraph::OnDown() {
...
+    }
+    
+    // Rollover conditions
+    // if the direction is up, rollover at the end of document
+    // if derction is down, rollover at start of document
+    
+    if (direction > 0) {
```
Comment (neurodroid) [inline src/stimfit/gui/graph.cpp]: These checks don't need to depend on `direction`

File: src/stimfit/gui/graph.cpp [hunk 4]
```diff
@@ -2152,6 +2163,50 @@ void wxStfGraph::OnDown() {
...
+	new_channel += direction;
+    }
+    
+    //----------------------------------------
+    // Update the window
+    //----------------------------------------
+    
+    // Pointer to wxStfChildFrame to access Channel selection combo
+    wxStfChildFrame* pFrame = (wxStfChildFrame*)Doc()->GetDocumentWindow();
+    if (!pFrame) {
```
Comment (AndrewMicallef) [inline src/stimfit/gui/graph.cpp]: Is this check even necessary?

File: src/stimfit/gui/graph.cpp [hunk 5]
```diff
@@ -2152,6 +2163,50 @@ void wxStfGraph::OnDown() {
...
+	new_channel = 0;
+    }
+    if (new_channel < 0) {
+	// Rollover to end if channel out of range
+	new_channel = Doc()->size() - 1;
+    } 
+    
+   if (new_channel == reference_ch) {
+	// Skip the reference channel
+	new_channel += direction;
```
Comment (AndrewMicallef) [inline src/stimfit/gui/graph.cpp]: Sorry, did this in a rush earlier today and didn't get a chance to compile / test.

Comment (AndrewMicallef) [inline src/stimfit/gui/graph.cpp]: 8389a3b worked... the refactorisation I just did breaks the rollover....

Comment (AndrewMicallef) [inline src/stimfit/gui/graph.cpp]: Is there like a debugging tool that I should be using for this? I can't seem to get back to a state with no errors and a refactored function.

File: src/stimfit/gui/graph.cpp (hunk 4 above)
Comment (neurodroid) [inline src/stimfit/gui/graph.cpp]: I don't think you can rule out that GetDocumentWindow() returns a null pointer.

File: src/stimfit/gui/graph.cpp (hunk 5 above)
Comment (neurodroid) [inline src/stimfit/gui/graph.cpp]: You can always check out a previous commit with git

Comment (neurodroid) [inline src/stimfit/gui/graph.cpp]: The problem here is that you're changing new_channel *after* you've done all the range checking - how can you rule out that new_channel is out of range after l2190?

File: src/stimfit/gui/graph.cpp [hunk 6]
```diff
@@ -2168,41 +2168,48 @@ void wxStfGraph::ChanScroll(int direction) {
...
-	new_channel = Doc()->size() - 1;
-    } 
     
-   if (new_channel == reference_ch) {
-	// Skip the reference channel
-	new_channel += direction;
+    if (new_chan > last_chan) {
+    	// Rollover to start if channel out of range
+    	// making sure to skip the reference channel
+        new_chan = (ref_chan == 0)? 1 : 0;
```
Comment (AndrewMicallef) [inline src/stimfit/gui/graph.cpp]: I hate ternary operators, but it made more sense to have a couple and do the reference channel check in line, than have a couple of nested `if`s

File: src/stimfit/gui/graph.cpp (hunk 5 above)
Comment (AndrewMicallef) [inline src/stimfit/gui/graph.cpp]: bc3d600 fixes this

File: src/stimfit/gui/graph.cpp (hunk 6 above)
Comment (neurodroid) [inline src/stimfit/gui/graph.cpp]: What happens for single-channel files in this case?

Comment (AndrewMicallef) [inline src/stimfit/gui/graph.cpp]: It'll probably throw an error if the default value for `Doc()->GetSecChIndex();`  is 0

I'll test it out once I've finished my recordings...

Comment (AndrewMicallef) [inline src/stimfit/gui/graph.cpp]: Although why would you should want to scroll a single channel file. It might be best to just test for this edge case and return early if `Doc->size() == 1`?

Comment (AndrewMicallef) [inline src/stimfit/gui/graph.cpp]: I'll fix that in about an hour

**Comment by AndrewMicallef** [comment]
- breaks out early now for the single channel case.
- Scrolling a two channel file does nothing; and I'm not sure if that is a good idea or not...
- Scrolling an N > 2 channel file shifts the active channel up or down by one;

File: src/stimfit/gui/graph.cpp [hunk 7]
```diff
@@ -2152,6 +2163,62 @@ void wxStfGraph::OnDown() {
...
+    */
+    
+    // direction is either +1, or -1 
+    
+    int ref_chan = Doc()->GetSecChIndex();  
+    int new_chan = Doc()->GetCurChIndex() + direction;
+    int last_chan = Doc()->size()-1;
+    
+    // Exit early if there is only one channel    
+    if (last_chan == 0) {
```
Comment (AndrewMicallef) [inline src/stimfit/gui/graph.cpp]: I use that variable, in 6 places...

**Comment by neurodroid** [comment]
Before we merge can you test your new shortcuts with
  * single channel files
  * 2 channel files
  * Using the shortcuts when no file is open at all

File: src/stimfit/gui/graph.cpp [hunk 8]
```diff
@@ -2175,7 +2175,7 @@ void wxStfGraph::ChanScroll(int direction) {
     int last_chan = Doc()->size()-1;
     
     // Exit early if there is only one channel    
-    if (last_chan == 0) {
+    if (Doc()->size() == 1) {
```
Comment (AndrewMicallef) [inline src/stimfit/gui/graph.cpp]: I'm doubt accessing this call twice in 2 lines is going to cause significant performance issues

Comment (neurodroid) [inline src/stimfit/gui/graph.cpp]: It's not about performance, didn't see that you were using `last_chan` elsewhere, both options are fine in this case

Comment (AndrewMicallef) [inline src/stimfit/gui/graph.cpp]: Just tested with files made from this script:
```{.python}
import stfio
import numpy as np


# construct Sections from arrays:
def seclist(N):
    return [stfio.Section(np.random.rand(5000))for i in range(N)]

    
def make_dummyfile(N_chans = 1):

    # construct Channels from lists of Sections
    chlist = [stfio.Channel(seclist(20)) for j in range(N_chans)]
    # Set channel units

    # construct a Recording from a list of channels
    rec = stfio.Recording(chlist)
    rec.dt = 0.05 # set sampling interval
    rec.xunits = "ms" # set time units

    return rec


make_dummyfile(N_chans = 1).write('D:/1_chan_test.h5');
make_dummyfile(N_chans = 2).write('D:/2_chan_test.h5');
make_dummyfile(N_chans = 20).write('D:/20_chan_test.h5');
```

Comment (AndrewMicallef) [inline src/stimfit/gui/graph.cpp]: - The no file case does nothing, unless the python terminal is selected, in which case it moves through the history.
- 1 channel : no effect
- 2 channel case : No effect
- 20 channels : works as expected

**Comment by neurodroid** [comment]
Merged, thanks for your contribution!

**Comment by AndrewMicallef** [comment]
Any time!
Does the manual now need a page for shortcuts / hotkeys?

**Comment by JoseGuzman** [comment]
it would be nice to document every new feature of Stimfit in the manual. In addition, you could add the new shortcuts to [this wikipage ](https://github.com/neurodroid/stimfit/wiki/Keyboard-shortcuts-for-Stimit)
